### ISSUE TITLE
JVM Support

### ISSUE THREAD

**Issue by murphye** [fsharp/fssf-ask-the-board#4]
# [Question/Proposal]: Title

JVM Support

# Description
[description]: #description

I believe there is room for another functional language to run on the JVM. I believe F# is a great language with limited reach on the .NET runtime.

# Motivation and impact
[motivation]: #motivation

F# will reach a much broader audience. It may also spur innovation with the language.

# Proposed solution and alternatives
[solution]: #solution

https://github.com/penberg/fjord is a previous effort to port F# to JVM.

# Drawbacks
[drawbacks]: #drawbacks

JVM will play second fiddle to .NET for F# support. Not having full support on equal level to .NET will be a problem.

# Are you willing to help?

I do not have expertise to help, sorry.

# Other thoughts
[other]: #other-thoughts

**Comment by realvictorprm** [comment]
If we can spend time on this we should rather go for an LLVM Backend. This would be of even more impact in my opinion.

**Comment by realvictorprm** [comment]
However in general providing as many backends as possible is still a good idea I think :smiley:. Thanks for your question @murphye!

**Comment by Krzysztof-Cieslak** [comment]
> If we can spend time on this we should rather go for an LLVM Backend

Who is 'we' ?

**Comment by realvictorprm** [comment]
Me and the rest of the working community @Krzysztof-Cieslak :wink:

**Comment by ReedCopsey** [comment]
This proposal is better suited as a language suggestion. I'd recommend filing under https://github.com/fsharp/fslang-suggestions if you want this to be discussed properly.

**Comment by dsyme** [comment]
@ReedCopsey I agree it's not really a thing for the FSSF board.  However it's also not really a language suggestion (where it would just gather dust)

It's really a proposal for a new community initiative to open up a whole new interpretation of F# in a new ecosystem.  Just like Fable, except for the JVM.

Just like Fable I'd be fully supportive of a community initiative in this area.  If you follow the pattern of Fable, and use the F# Compiler Service as the core of your compiler, then it is actually possible to get quite a long way quickly.  

So, the question is - is anyone game to take this on and try to produce a variant of F# for the JVM, based on FCS? It's a non-trivial commitment and [can lead you a long way...](https://www.eventbrite.es/e/fableconf-bordeaux-tickets-34089709238)

**Comment by dsyme** [comment]
Re LLVM - F# already has an LLVM backend by virtue of Mono. I don't know the technical details, but Mono certainly produces LLVM code.  

However AFAIK LLVM doesn't define an interop standard beyond C calls, so it's different to the JVM or Javascript, since it's not an ecosystem of interoperating code, libraries, tools and methodologies.  It's just a native code generator.

**Comment by realvictorprm** [comment]
@dsyme this still sounds interesting. 

Personally I'm interested in using F# in the native environment.

**Comment by lenadroid** [comment]
Thanks @murphye for bringing this up to the Boards attention! Thanks to everyone who commented on this issue, have in mind that we aim to keep these discussions positive.

This issue has been discussed at the Board meeting on Monday 8/7/2017 and Trustees believe that this idea might have future as an independent OSS F# Community project given the support and involvement of F# Community. 

Therefore, we would definitely like to give this issue more visibility in the F# Community. There will be a tweet from fsharp.org account and link posted in FSSF slack channel to attract volunteers and inspire discussion.

Please, feel free to share this issue on Twitter, [FSSF slack](https://fsharp.slack.com/) #compiler channel ([join](http://foundation.fsharp.org/join) FSSF for invite) or other communication channels to discuss and attract people who might be interested in being involved. 

This issue will be closed in 2 weeks, as it is out of the scope of the Foundation and its [mission](http://foundation.fsharp.org/).

**Comment by alexandru** [comment]
Hi all, 

I'm a Scala developer and I played with F# as well.
I've seen the tweet and thought of giving my 2 cents.

I would like to see F# on top of the JVM. The JVM ecosystem thrives on diversity, multiple languages coexist within it and have quite big communities, including Scala, Clojure, Kotlin and JRuby and even if the groups of developers and companies using these tend to be distinct, there's a lot of sharing of knowledge, libraries and tools. And apparently we now have Haskell too, see [Eta-lang.org](http://eta-lang.org/), including an SBT plugin for [integrating it in Scala projects](https://blog.eta-lang.org/integrating-eta-into-your-scala-projects-a8d494a2c5b0).

F# is a cool language and personally I love the idea of targeting multiple platforms with the same language, but its success depends on how this vision is going to be executed.

For example we have [Scala.js](http://www.scala-js.org/), the Scala compiler for JavaScript / Node.js, which is a big success. But besides the excellent implementation, Scala.js is a success because:

1. it shares all of Scala's tooling, it reuses the compiler's frontend, it uses Scala's [SBT](http://www.scala-sbt.org/) as the official build tool, it uses Maven Central as the official way to distribute dependencies (N.B. not NPM), IDEs work with it out of the box, etc.
2. the community rallied around it and now most popular libraries that can be cross-compiled to Scala.js are cross-compiled, such that people can use the exact same libraries, with minor differences where needed (e.g. obviously on top of JS you can't block threads)

I'm mentioning thinking of past attempts I've seen for an F# to JS compiler. If there isn't sharing somehow (e.g. cross-compiled libraries), then personally I think it's tough to gain traction.

Btw, we now have [Scala Native](http://www.scala-native.org/) as well. It has slightly different semantics here and there. Cool project, following the same recipe of Scala.js, although it's harder to bootstrap. For example Scala for the JVM and Scala.js didn't have to provide a Garbage Collector that doesn't suck, or worry about the object model.

So I would love to see F# on top of the JVM, I think it has great potential, but there are going to be challenges. The barrier to entry is a little high, but once polished, I'm pretty sure that it will have adopters, because the ecosystem is big and we aren't waiting on any single company to bless technologies :-P

**Comment by zpodlovics** [comment]
As far as I know the closest ML thing you can have now is Xavier Clerc OcamlJava ( http://ocamljava.org , https://github.com/xclerc/ocamljava ) that's already run on the JVM. 

I am afraid it's not that simple to move a language to an another platform with different underlying semantics. How the F# language is supposed to work on top of JVM? How would you translate and map the F# language semantics and rules to JVM? What about the memory model (-> JMM) changes? What about the data representation changes (no unsigned integers, no value types (yet)) ? How would you do tailcalls (not even Mono figured out correctly: https://bugzilla.xamarin.com/show_bug.cgi?id=12635#c11)? How do you make sure the implementation is complete and correct?

Take a look at for example Denis Bogdanas, Grigore Rosu: K-Java: A Complete Semantics of Java
http://fsl.cs.illinois.edu/index.php/K-Java:_A_Complete_Semantics_of_Java
http://fslweb.cs.illinois.edu/FSL/papers/2015/bogdanas-rosu-2015-popl/bogdanas-rosu-2015-popl-public.pdf
http://fslweb.cs.illinois.edu/FSL/presentations/2015/2015-01-16-K-Java-POPL/2015-01-16-K-Java-POPL.pdf

They defined Java 1.4 with ~1000 semantics rules to map Java language semantics to JVM semantics. Without these complete set of semantics rules, how would you write the F# compiler/transpiper to JVM? GraalVM could be an even better research target.

Please note: with some weird twists you can already run LLVM code (which probably include the Mono case too) on JVM using GraalVM + Sulong:
https://github.com/graalvm/sulong
https://www.youtube.com/watch?v=yyDD_KRdQQU

@realvictorprm If you want to experiment with F# and LLVM, take a look (shameless plug) at FsHlvm which is a DSL like high level virtual machine on top of LLVM with generics & type specialization, garbage collection, multi-core, ffi, etc. It was recently updated to .NET Core 2.0 Preview2 and LLVM 3.8 (Ubuntu 16.04 x86_64): 
https://github.com/kp-tech/fshlvm

I will do some experiment (when I have enough time) in the future with porting the F# Quotation Compiler to FsHlvm. However I am afraid, there will be no correct and complete solution without figuring out the sematics translation rules and mapping from F# to LLVM (https://github.com/kframework/llvm-semantics).

**Comment by realvictorprm** [comment]
@dsyme how do you think about the technical situation? Did the situation change compared to the one three years ago?

**Comment by alfonsogarciacaro** [comment]
Hi there! I'll try to give my 2 cents about porting F# to other platforms. Please note these are personal opinions so they may be wrong :)

### About porting F# to a new platform

Adding a new compilation target to F# is considerable work so before starting it it's better to assess what the potential adoption could be. When devs consider using an alternative language instead of the platform's native one, they weigh:

- advantages of using the language (usually because they like its features and tooling), against
- friction of using a _foreign_ language (usually in terms of performance and when interoperating with native code)

Because of this, again in my opinion, just compiling F# to another platform is useless if there's no good integration with the new ecosystem. This is what I tried to do with Fable from the beginning even at the cost of little sacrifices in F# semantics and a few controversial decisions (like erasing options). In other areas, like reflection I've tried to find a balance between allowing the most common operations without adding any overhead to the generated code.

### Effort to port F# to the JVM

Although I'm not a JVM expert, I don't think I'm wrong to say that porting F# to the JVM will be a huge effort. Although I'm the first to praise the magic results of non-directed collaboration in the F# (or many others) community, I'm with @Krzysztof-Cieslak in that something like this won't be done by an abstract _we_, it requires the compromise of one or two people to dedicate the time need to create the first prototype and from there iterate with the feedback from early users.

If you see the [source code of Fable](https://github.com/fable-compiler/Fable/tree/master/src/dotnet/Fable.Compiler) you'll see there're two steps in the compilation: first the AST provided by FSharp.Compiler.Service is translated to Fable's specific AST, and then Fable AST is translated to [Babel's counterpart](https://github.com/babel/babylon/blob/master/ast/spec.md). The reason to have this intermediate AST is to have something easier to manage for replacements (calls to .NET BCL mainly) and plugins. But it was also designed to make it easier to target other platforms with Fable. Actually at the very beginning I was talking to another user about the possibility of using Fable to compile to the JVM but this didn't happen at the end.

So far I've been only caring about JS when evolving the Fable AST but if somebody else wants to use it to compile F# to the JVM it shouldn't be too difficult to polish it so it can be used for any target with OOP basics (classes, simple inheritance) and lambda support (please note I cannot compromise myself to anything beyond that). I think Fable AST is valuable because it already removes several FCS quirks, simplifies things like pattern matching and optimizes common uses like pipes. Also ideally, a 3rd party tool to generate JVM bytecode from an AST should be used the same way Fable uses Babel.

As a side note, there's recent work to port F# to BEAM (Erlang's VM) also using FCS but not related to Fable in any way. It's probably worth a look (I forgot the repo name sorry!) but note BEAM's semantics seem to be much more different than OOP.

### Make F# easier to port to other platforms

I'll finish mentioning a couple of things that have already been discussed in other threads and would make F# easier to port to other platforms. Ideally these changes should be done both in FCS and the IDE.

- Make it possible to start F# projects referencing a minimal mscorlib/FSharp.Core combo. Most of difficulties user have when writing F# for other platforms is they don't exactly know which APIs are supported.
- Create flags for some F# features so the user knows which things are not available when targeting other platforms (e.g. in JS you cannot pass numbers by reference).

**Comment by alfonsogarciacaro** [comment]
@alexandru If you're interested in using F# for JS, you may want to have a look at [Fable](http://fable.io). I don't know Scala.js in depth but I've talked to the author and in my understanding Scala.js does try to port the full Scala semantics to JS even at the cost of some overhead. This makes it much easier to port libraries to Scala.js, than it's for Fable (though in F# case one of the main issues is libraries make use of many APIs of the huge BCL, something that's been affecting the creation of cross-platform libraries in .NET for years).

**Comment by shishini** [comment]
I think F# on JVM, will be a distraction

The JVM have, so many languages, Clojure and Scala probably occupy the mind-share of those who would be interested in F#

F#, is also a distraction, because there is already Ocaml and Reason 

F# key selling point, is that it is a .Net language, the objective should be to make it the number 1 .Net language, for F# to surpass C# and VB .Net

Make sure that F# can be used in every use case where C# is used

**Comment by kjnilsson** [comment]
For completeness the fsharp to beam port mentioned by @alfonsogarciacaro is [fez](https://github.com/kjnilsson/fez). 

Although fez uses FCS and translate the TAST just like Fable it is different in that it compiles to a much more restricted target (core erlang) which means certain features (like mutation) may never be available. Hence there are real limits to what the project can reasonably achieve. (I do think something useful can come out of it but it is a different discussion). 

What I have noticed is how closely tied fsharp is to .NET.  The structure of the typed AST for example doesn't feel like it was ever meant to be compiled to anything else than IL (which is probably true). Unless this changes I don't think achieving the full fsharp experience on another platform is possible as a community effort. With commercial backing it could be possible, perhaps but that is unlikely given there are so many options on the jvm.

I would like to see an intermediate representation between the Untyped and Typed AST which is independent of compilation target. Like a more "functional" TAST similar to how the erlang compilation pipeline was changed to include [core erlang](https://www.it.uu.se/research/group/hipe/cerl/) - a simple, machine manipulable language that works as an intermediate between the erlang AST and the beam representation. This would then provide an obvious higher-level place to hook other backend in at the language level at least.

One big question is what to do about `System.*`? Is fsharp intrinsically linked to the .NET framework? How much of `System.*` should one need to translate? With fez I early made the decision _not_ to translate anything from `System.*` - this proved difficult again because so much is coupled to .NET so I've had to relax that approach somewhat. The question is where do you draw the lines? And once you've drawn the lines is it still fsharp you are left with or something else?

**Comment by alfonsogarciacaro** [comment]
Thanks for the link @kjnilsson! Yes, I had the same feeling working with FCS. Fable tries to fix that for the AST (even if you can also find some IL there) converting it into something more neutral, though it resembles more an OOP language with lambda support (like ES2015 or Java 8) than a functional one.

Again, for me the main issue is the assumption all F# projects references the whole BCL/FSharp.Core, if it were possible to count only on a minimal API, it'd be much easier to create a prototype of F# compiling to other languages with basic support, and iterate from there.

The .dll reference issue is also a technical problem for Fable's [online REPL](http://fable.io/repl.html): even if FCS+Fable have been translated to JS and can be run in the browser frontend, users still have to [download many assemblies](https://github.com/fable-compiler/Fable/blob/40ab86cb364fdb3246735bcab17a272d0805c573/src/dotnet/Fable.JS/demo/worker.js#L27-L38) to be able to parse a single line of F# code.

Also, if some F# features could be disabled through compiler flags (like mutability) I guess it'd much easier the F# subset that can be compiled to BEAM in your case.

**Comment by OnurGumus** [comment]
How about the lack of reified generics in jvm?

**Comment by s952163** [comment]
Besides, the technical feasibility, who and why (e.g. what problem/need does it solve) would use F# on the JVM? Scala died on .NET, and I don't see a different fate for F# on the JVM. For example Scala is being actively developed to take advantage of the Java versions (e.g. incorporating lambdas into its byte code generation from Java 8). Will F# be maintained/developed at/with the same pace/effort? Scala already fuses OOP and FP on the JVM quite well (some might say better then F# on .NET). F# has some advantage in terms of syntax but that is usually not enough. Its main selling point is to provide FP on .NET, which it does quite well, albeit being limited by the .NET ecosystem itself. Which is just trying to do what JVM has done for what, over 20 years, i.e. being cross-platform....

**Comment by kspeakman** [comment]
I recently went looking for F# on JVM, so I guess you could say that I am in the target market for such a thing. Basically, I was looking at alternatives to F# on .NET. I considered a number of options such as Scala or Clojure on JVM, Erlang/Elixir on BEAM, OCaml, Haskell, various things on Node.js. OCaml is probably closest language-wise, but has a number of tooling / ecosystem issues for our purposes. JVM probably has the largest ecosystem and tooling selection, but lacks functional languages I actually want to program in. (Clojure I want to like, but I don't care for the tradeoff that comes with being untyped.)

At the end of the day, F# strikes a great balance (examples: performance vs OCaml, approachability vs Haskell, signal-to-noise vs Scala, etc) that I haven't been able to find elsewhere yet. Aside from the general reasons of JVM having broader reach/ecosystem, another motivation I have to be off Microsoft stack is their insistence on opt-out (or in some cases impossible to turn off) data collection in every product, now [including build tools](https://github.com/dotnet/cli/issues/3093). For now F# is keeping me on .NET. But I would love to see F# not be limited to one platform, as it is a great language with well-chosen tradeoffs IMO.

**Comment by realvictorprm** [comment]
As told by @lenadroid this is going to be closed after 2 weeks (already a bit more). 
Thank you all for the discussions!

**Comment by 21c-HK** [comment]
Just throwing out a few reasons why efficient F# support on the JVM will be difficult, because the JVM (while apparently having a better garbage collector) has limited support for writing efficient code:

- missing support for proper / reified / non-erased generics
- missing support for user-defined value types
- missing support for tail call optimization
- missing support for non-virtual methods
- missing support for explicitly inlining functions
- missing support for call-by-reference
- missing support for pointers
- incredibly slow FFI (apparently 10x slower than .NET FFI according to Jon Harrop)

**Comment by kspeakman** [comment]
@21c-HK You also forgot mutually recursive functions. But picking on the JVM's chose tradeoffs isn't helpful to this suggestion. Especially considering other functional languages already exist which support (albeit limited) TCO and more typing features than F# (HKTs, although still missing Union Types???). Granted _I_ don't want to write in Scala, but some pretty interesting things have been made with it nonetheless.

**Comment by alexandru** [comment]
Scala can express logical disjunctions, although with a weird encoding (e.g. the data constructors are defined as `case class` extending from a `sealed` trait or class) and then the compiler can do exhaustiveness checks in pattern matching, just like F#. In [Dotty](http://dotty.epfl.ch/) they introduced `enum` which makes this official, along with union types (e.g. `A | B`).

The lack of TCO support in the runtime is regrettable, however OpenJDK is a lively project which has promising experiments being worked on, see for example Project Loom, which should include TCO: [announcement](http://mail.openjdk.java.net/pipermail/discuss/2017-September/004390.html) and [proposal](http://cr.openjdk.java.net/~rpressler/loom/Loom-Proposal.html).

That said, we have had libraries and abstractions inspired by Haskell for dealing with tail-recursive functions.

For example Monix's [Task](https://monix.io/docs/2x/eval/task.html) and [Coeval](https://monix.io/docs/2x/eval/coeval.html), inspired by Scalaz and by Haskell's `IO`, are powered by very efficient trampolines and can be used to express mutually recursive functions. If interested, see [my presentation from Scala Days](https://www.youtube.com/watch?v=wi97X8_JQUk).

Also the shinning star of the JVM is currently [Eta](http://eta-lang.org/), a full fledged and commercially supported Haskell implementation — because their interop happens through a FFI, they do not need to preserve stack traces or rely on Java's stack too much, so they can afford optimizations that are not afforded by libraries, like being more aggressive in transforming recursions into loops and generating data type specific trampolines and I quote from their [FAQ](http://eta-lang.org/docs/html/faq.html):

> We don’t do anything to handle direct tail calls per se. At compile-time, we optimize tail recursive-calls into loops and we use the fact that lazy evaluation controls stack growth, unless you’re using too much of it. You can think of lazy evaluation as a hierachy of trampolines, but unlike general-purpose trampolines, each “trampoline” is specialized to a given type. And instead of executing a unknown virtual method call after getting a closure, it checks the tag of the closure which will help it select a branch to execute next.

The results are really promising and I quote:

> Benchmarks show that Eta can be within the performance of GHC in a few cases and within 2x in many other cases after JIT optimizations. 

This is quite an achievement given that they're just getting started and given that many years of research have gone into GHC to optimize it for Haskell's specific workload. Make no mistake, GHC is among the best, if not the best compiler and runtime for a purely FP language, even if it doesn't have the backing of a big corporation. So to see Eta being within 2x of its performance, given that Haskell is lazy and filled with recursions, that's quite amazing.

But more to the points raised by @21c-HK, because I love a good vs comparison 😀

> missing support for proper / reified / non-erased generics

There is no "proper". All that "reified generics" are doing is to add type info at runtime, however you don't need reification with a good enough static type system. You rarely need to do `isInstanceOf` checks in Scala for example and that's only when you're dealing with Java-style code and you never need to do that in Haskell, a language where everything is solved at compile time, including polymorphic call sites.

What's missing in Java is actually *specialization*, but — (1) those are coming in [Project Valhalla](http://cr.openjdk.java.net/~briangoetz/valhalla/specialization.html) and (2) they are not such a good idea for the proliferation of languages, I hope they handle it well, because specialization can also be achieved at compile time, with no runtime support and adding that to the runtime can turn out to be mission impossible without crippling it for alternative languages.

Anyway, there have been projects for specialization in Scala, including the `@specialized` annotation included in the language, but that has a less fortunate implementation, along with [Miniboxing](http://scala-miniboxing.org/), a later and much better attempt. But the efforts have now moved to Dotty, animated by the upcoming TASTY binary format, which should make recompilation easier.

But going back to reification, that actually quite bad news for Higher-Kinded Types, because if you want HKTs, you now need to add it to the runtime, otherwise risk interop problems, along with second-class citizenship. This is why F# won't do it, even though HKTs are very important for expressing many classes of abstractions in FP.

People used to say that .NET supports any language as long as it is C#. And this is precisely what they meant 😉

> - missing support for explicitly inlining functions

@21c-HK FYI that's a function of the compiler, not the runtime, but since you mention it, the JVM is among the few mainstream virtual machines that do inlining of virtual method calls and it does so based on runtime introspection. That's what the *Hotspot VM* introduced in Java 1.4 is all about.

It's so good actually that if you do the inlining by yourself, you're actually making it more difficult for the runtime to apply optimizations. And Scala does have the ability to [mark functions for inlining](https://www.scala-lang.org/api/current/scala/inline.html), the surprise for beginners being that many times the "optimized" code is slower.

On the other points:

> - missing support for non-virtual methods

This isn't true of course and the runtime has several invocation instructions, depending on the method's type, i.e. static, final, etc.

> - missing support for pointers
> - incredibly slow FFI (apparently 10x slower than .NET FFI according to Jon Harrop)

These items are irrelevant, plus with all due respect, but I would take Jon Harrop's claims with a grain of salt, even though I usually love reading his rants.

Cheers,

**Comment by ShalokShalom** [comment]
@kspeakman You may get interested in Fantom: http://fantom.org/doc/docLang/Functions

**Comment by ShalokShalom** [comment]
Hnn, I think there is a pretty usable way, to bring F# on the JVM.
I might oversee something, while I think that two things combined can reach that goal with a reasonable effort and great side-effects. 

Correct me, if I am wrong: FSharp > Ocaml > [OCaml-Java](https://discuss.ocaml.org/t/ocaml-java-compiler/1450)

Of course does this come with things like the maintenance for OCaml-Java and this FSharp to OCaml compiler while I still think this means a proper way for a lot of scenarios to run FSharp code on the JVM and a source to source translation to OCaml is worthfully anyway.

All the tools could be used and one of the next logical steps is to list the syntactic differences between OCaml and FSharp. 

Which is again a nice side effect since many people wonder about these while this fundamental question seems unanswered?

**Comment by jkone27** [comment]
what about Fable (now python support was added in alpha)? https://github.com/fable-compiler/Fable.Python , could a Java transpilation process be possible in theory

**Comment by ReedCopsey** [comment]
@jkone27 I'd ask on the Fable repositories if you're truly interested, but this would also require somebody being willing and able to put in that effort.
